### PR TITLE
OP-438 New HTTP Client for SMS Gateway

### PR DESCRIPTION
**Pull request by AndreiDodu**
refactoring (added feign) + added textbelt gateway

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -18,6 +19,9 @@
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<spring.framework.version>4.3.28.RELEASE</spring.framework.version>
 		<license.maven.plugin.version>4.1</license.maven.plugin.version>
+		<!-- <spring-cloud-starter-openfeign>2.0.0.RELEASE</spring-cloud-starter-openfeign> -->
+		<spring-cloud-starter-openfeign>1.4.7.RELEASE</spring-cloud-starter-openfeign>
+		<!-- <spring-cloud-starter-netflix-ribbon>2.0.0.RELEASE</spring-cloud-starter-netflix-ribbon> -->
```
Comment (mwithi) [inline pom.xml:24]: should we keep these comments?

File: pom.xml [hunk 2]
```diff
@@ -82,25 +86,12 @@
...
-						</configuration>
-					</execution>
-				</executions>
-			</plugin> -->
+			<!-- <plugin> <groupId>org.apache.maven.plugins</groupId> <artifactId>maven-dependency-plugin</artifactId> 
+				<executions> <execution> <id>copy-dependencies</id> <phase>prepare-package</phase> 
+				<goals> <goal>copy-dependencies</goal> </goals> <configuration> <outputDirectory>${project.build.directory}/OpenHospital20/lib</outputDirectory> 
+				<overWriteReleases>false</overWriteReleases> <overWriteSnapshots>false</overWriteSnapshots> 
+				<overWriteIfNewer>true</overWriteIfNewer> </configuration> </execution> </executions> 
+				</plugin> -->
```
Comment (mwithi) [inline pom.xml:94]: please rollback unnecessary changes

File: pom.xml [hunk 3]
```diff
@@ -199,15 +190,15 @@
...
-		    </exclusions>
+			<groupId>net.sf.jasperreports</groupId>
+			<artifactId>jasperreports</artifactId>
+			<version>6.16.0</version>
+			<exclusions>
+				<exclusion>
+					<groupId>com.lowagie</groupId>
+					<artifactId>itext</artifactId>
+				</exclusion>
+			</exclusions>
```
Comment (mwithi) [inline pom.xml:201]: the correct version is 6.17 as latest bump

File: pom.xml [hunk 4]
```diff
@@ -409,39 +400,91 @@
...
-		<!-- <repository>
-			<id>dcm4che</id>
-			<name>Dcm4Che Repository</name>
-			<url>https://www.dcm4che.org/maven2/</url>
-			<snapshots>
-				<enabled>false</enabled>
-			</snapshots>
-		</repository> -->
+		<!-- <repository> <id>dcm4che</id> <name>Dcm4Che Repository</name> <url>https://www.dcm4che.org/maven2/</url> 
+			<snapshots> <enabled>false</enabled> </snapshots> </repository> -->
```
Comment (mwithi) [inline pom.xml:487]: please rollback unnecessary changes

File: src/main/java/org/isf/generaldata/SmsParameters.java [hunk 5]
```diff
@@ -22,45 +22,44 @@
...
+
+	private static SmsParameters mySingleData;
+
+	private SmsParameters(String fileProperties) {
+		super(fileProperties);
+
+		TIMEOUT = myGetProperty("sms.gateway.thread.timeout", DEFAULT_TIMEOUT);
+		LOOP = myGetProperty("sms.gateway.thread.loop", DEFAULT_LOOP);
+		ICC = myGetProperty("sms.gateway.thread.icc", DEFAULT_ICC);
+		System.err.println(ICC);
```
Comment (mwithi) [inline src/main/java/org/isf/generaldata/SmsParameters.java:51]: please remove this or use LOGGER

File: src/main/java/org/isf/sms/service/SmsSender.java [hunk 6]
```diff
@@ -47,7 +46,7 @@ public SmsSender() {
 		LOGGER.info("SMS Sender started...");
 		SmsParameters.initialize();
 		delay = SmsParameters.LOOP;
-		LOGGER.info("SMS Sender loop set to {} seconds.", delay);
+		LOGGER.info("SMS Sender loop set to {} seconds.", Integer.valueOf(delay));
```
Comment (mwithi) [inline src/main/java/org/isf/sms/service/SmsSender.java:49]: maybe unnecessary boxing

File: src/main/java/org/isf/sms/service/SmsSender.java [hunk 7]
```diff
@@ -57,77 +56,50 @@ public void run() {
 			List<Sms> smsList = null;
 			try {
 				smsList = smsOp.getList();
-			} catch (OHServiceException ohServiceException) {
-				LOGGER.error(ohServiceException.getMessage(), ohServiceException);
+			} catch (OHServiceException e1) {
+				e1.printStackTrace();
```
Comment (mwithi) [inline src/main/java/org/isf/sms/service/SmsSender.java:60]: the correct form is with `ohServiceException` and `LOGGER.error`

File: src/main/java/org/isf/sms/service/SmsSender.java [hunk 8]
```diff
@@ -57,77 +56,50 @@ public void run() {
...
-									sms.setSmsDateSent(new Date());
-									try {
-										smsOp.saveOrUpdate(sms);
-									} catch (OHServiceException ohServiceException) {
-										LOGGER.error(ohServiceException.getMessage(), ohServiceException);
-									}
-									LOGGER.debug("Sent");
-								} else {
-									LOGGER.error("Not sent");
+				LOGGER.info("Found {} SMS to send", Integer.valueOf(smsList.size()));
```
Comment (mwithi) [inline src/main/java/org/isf/sms/service/SmsSender.java:63]: unnecessary boxing

File: src/main/java/org/isf/sms/service/SmsSender.java [hunk 9]
```diff
@@ -57,77 +56,50 @@ public void run() {
...
+				if (sender.initialize()) {
+					for (Sms sms : smsList) {
+						if (sms.getSmsDateSched().before(new Date())) {
+							boolean result = sender.sendSMS(sms);
+							if (result) {
+								sms.setSmsDateSent(new Date());
+								try {
+									smsOp.saveOrUpdate(sms);
+								} catch (OHServiceException e) {
+									e.printStackTrace();
```
Comment (mwithi) [inline src/main/java/org/isf/sms/service/SmsSender.java:74]: the correct form is with `ohServiceException` and `LOGGER.error`

File: src/main/java/org/isf/sms/service/SmsSender.java [hunk 10]
```diff
@@ -57,77 +56,50 @@ public void run() {
...
 				}
 			} else {
 				LOGGER.debug("No SMS to send.");
 			}
 			try {
 				Thread.sleep(delay * 1000);
-			} catch (InterruptedException interruptedException) {
-				LOGGER.error(interruptedException.getMessage(), interruptedException);
+			} catch (InterruptedException e) {
+				e.printStackTrace();
```
Comment (mwithi) [inline src/main/java/org/isf/sms/service/SmsSender.java:95]: the correct form is with `interruptedException` and `LOGGER.error`

File: src/main/java/org/isf/sms/service/SmsSender.java [hunk 11]
```diff
@@ -57,77 +56,50 @@ public void run() {
...
 			}
 		}
 	}
 
 	/**
-	 * @param running the running to set
+	 * @param running
+	 *            the running to set
```
Comment (mwithi) [inline src/main/java/org/isf/sms/service/SmsSender.java:102]: unnecessary new line

File: src/main/resources/applicationContext.xml [hunk 12]
```diff
@@ -53,31 +65,35 @@
...
+			<bean
+				class="org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter" />
+		</property>
+		<property name="jpaProperties">
+			<props>
+				<prop key="hibernate.connection.zeroDateTimeBehavior">convertToNull</prop>
+				<prop key="hibernate.show_sql">${hibernate.show_sql:false}</prop>
+				<prop key="hibernate.format_sql">${hibernate.format_sql:true}</prop>
+				<prop key="hibernate.dialect">${hibernate.dialect:org.hibernate.dialect.MySQLDialect}
+				</prop>
```
Comment (mwithi) [inline src/main/resources/applicationContext.xml:83]: unnecessary new line

File: src/main/resources/applicationContext.xml [hunk 13]
```diff
@@ -53,31 +65,35 @@
...
- 				<prop key="hibernate.dialect">${hibernate.dialect:org.hibernate.dialect.MySQLDialect}</prop>
- 				<prop key="hibernate.id.new_generator_mappings">false</prop>
+
+	<bean id="entityManagerFactory"
+		class="org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean">
+		<property name="dataSource" ref="springDataSource" />
+		<property name="packagesToScan" value="org.isf" />
+		<property name="jpaVendorAdapter">
+			<bean
+				class="org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter" />
```
Comment (mwithi) [inline src/main/resources/applicationContext.xml:75]: unnecessary new line

File: src/main/resources/applicationContext.xml [hunk 14]
```diff
@@ -53,31 +65,35 @@
...
+			</props>
+		</property>
+	</bean>
+
+	<tx:annotation-driven />
+	<bean id="transactionManager"
+		class="org.springframework.orm.jpa.JpaTransactionManager">
+		<property name="entityManagerFactory"
+			ref="entityManagerFactory" />
+	</bean>
```
Comment (mwithi) [inline src/main/resources/applicationContext.xml:95]: unnecessary new lines

**Review by mwithi — COMMENTED** [review]
I've set global `sms.gateway.thread.icc` and I've tried default token `textbelt` (1 free SMS/Day) but it seems not sending

**Comment by mwithi** [comment]
image

File: src/main/java/org/isf/sms/model/Sms.java [hunk 15]
```diff
@@ -1,23 +1,24 @@
 /*
- * Open Hospital (www.open-hospital.org)
- * Copyright © 2006-2020 Informatici Senza Frontiere (info@informaticisenzafrontiere.org)
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/sms/model/Sms.java:3]: Why change the default license block that is generated by a maven target?  Please rollback.

File: src/main/java/org/isf/sms/providers/SmsSenderInterface.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/sms/providers/SmsSenderInterface.java:1]: As noted above this is not the current license header.

File: src/main/java/org/isf/sms/service/SmsSender.java [hunk 17]
```diff
@@ -57,77 +56,50 @@ public void run() {
 			List<Sms> smsList = null;
 			try {
 				smsList = smsOp.getList();
-			} catch (OHServiceException ohServiceException) {
-				LOGGER.error(ohServiceException.getMessage(), ohServiceException);
+			} catch (OHServiceException e1) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/sms/service/SmsSender.java:59]: Logging exceptions with message and trace are the standard; not just printStackTrace()

File: pom.xml (hunk 1 above)
Comment (AndreiDodu) [reply pom.xml]: No, I deleted them.

File: pom.xml (hunk 2 above)
Comment (AndreiDodu) [reply pom.xml]: ok, done.

File: pom.xml (hunk 3 above)
Comment (AndreiDodu) [reply pom.xml]: updated the current version

File: pom.xml (hunk 4 above)
Comment (AndreiDodu) [reply pom.xml]: done

File: src/main/java/org/isf/generaldata/SmsParameters.java (hunk 5 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/generaldata/SmsParameters.java]: done

File: src/main/java/org/isf/sms/service/SmsSender.java (hunk 10 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/sms/service/SmsSender.java]: done

File: src/main/java/org/isf/sms/service/SmsSender.java (hunk 11 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/sms/service/SmsSender.java]: done

File: src/main/resources/applicationContext.xml (hunk 12 above)
Comment (AndreiDodu) [reply src/main/resources/applicationContext.xml]: corrected

File: src/main/resources/applicationContext.xml (hunk 13 above)
Comment (AndreiDodu) [reply src/main/resources/applicationContext.xml]: corrected

File: src/main/resources/applicationContext.xml (hunk 14 above)
Comment (AndreiDodu) [reply src/main/resources/applicationContext.xml]: corrected

File: src/main/java/org/isf/sms/model/Sms.java (hunk 15 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/sms/model/Sms.java]: restored

File: src/main/java/org/isf/sms/providers/SmsSenderInterface.java (hunk 16 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/sms/providers/SmsSenderInterface.java]: done

File: src/main/java/org/isf/sms/service/SmsSender.java (hunk 6 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/sms/service/SmsSender.java:49]: seems that it wants an object

File: src/main/java/org/isf/sms/service/SmsSender.java (hunk 8 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/sms/service/SmsSender.java:63]: also here it wants an object.

File: src/main/java/org/isf/sms/service/SmsSender.java (hunk 7 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/sms/service/SmsSender.java:60]: corrected

File: src/main/java/org/isf/sms/service/SmsSender.java (hunk 9 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/sms/service/SmsSender.java:74]: corrected

File: src/main/java/org/isf/sms/providers/skebby/SkebbyGatewayService.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+		String baseUrl = this.smsProperties.getProperty(this.getRootKey() + ".ribbon.base-url");
+		return Feign.builder().encoder(new CustomCommonEncoder()).decoder(new CustomCommonDecoder()).logger(new Slf4jLogger(SkebbyGatewayRemoteService.class))
+						.logLevel(feign.Logger.Level.FULL).contract(new SpringMvcContract()).target(SkebbyGatewayRemoteService.class, baseUrl);
+	}
+
+	private String loginUserKeySessionKey() {
+		// USER_KEY and ACCESS_TOKEN avoids the login call every time we need to send sms
+		final String userKey = this.smsProperties.getProperty(KEY_USER_KEY);
+		final String token = this.smsProperties.getProperty(KEY_ACCESS_TOKEN);
+		if (userKey != null && !userKey.isBlank() && token != null && !token.isBlank()) {
```
Comment (mwithi) [inline src/main/java/org/isf/sms/providers/skebby/SkebbyGatewayService.java:113]: isBlank has been introduced with JDK11 so the maven build fails for JDK8.
Maybe we can replace with .trim().isEmpty()?

**Review by mwithi — CHANGES_REQUESTED** [review]
Maven build fails for JDK8 because of String.isBlank() method

File: src/main/java/org/isf/sms/providers/skebby/SkebbyGatewayService.java (hunk 18 above)
Comment (mwithi) [reply src/main/java/org/isf/sms/providers/skebby/SkebbyGatewayService.java:113]: Done locally, will be pushed with the merge commit

**Comment by mwithi** [comment]
Despite the test sms seems temporary disabled it seems working fine in the latest version!
image

**Comment by mwithi** [comment]
it works like a sharm! 👏👏 I'm going to merge this locally adding the .isBlank() substitution for JDK8 compliance.
Thanks!
